pkg/pace.go

```diff
@@ -63,14 +75,39 @@ func (p *pacer) onPhaseElapsed() {
 }
 
 func (p *pacer) waitDuration(t time.Time) time.Duration {
+	now := time.Now()
+
 	// If there are no more phases left, continue with the last phase's rate
 	if p.lastRequestTime.IsZero() {
 		p.lastRequestTime = t
+		p.lastRequestWallTime = now
+		p.phaseStartRequestTime = p.lastRequestTime
+		p.phaseStartWallTime = p.lastRequestWallTime
+	}
+
+	originalDurationFromPhaseStart := t.Sub(p.phaseStartRequestTime)
+	expectedDurationFromPhaseStart := time.Duration(float64(originalDurationFromPhaseStart) / p.phases[0].rate)
+	expectedWallTime := p.phaseStartWallTime.Add(expectedDurationFromPhaseStart)
+
+	if p.ReportInterval > 0 {
```

Comment:
@szmglh maybe extract a function for this, as the current one is getting a bit long and is already "dense" because it contains a critical section of the main algo.

Reply:
Sure, done in 470f8bcfef82a201fd49c67220af64df43b24b45